pkg/util/virtualmachine/virtualmachine.go

```diff
@@ -0,0 +1,53 @@
+package virtualmachine
+
+import (
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+
+	ctlkubevirtv1 "github.com/harvester/harvester/pkg/generated/controllers/kubevirt.io/v1"
+)
+
+// IsVMStopped checks VM is stopped or not. It will check two cases
+// 1. VM is stopped from GUI
+// 2. VM is stopped inside VM
+// These two cases are all stopped case.
+func IsVMStopped(
+	vm *kubevirtv1.VirtualMachine,
+	vmiCache ctlkubevirtv1.VirtualMachineInstanceCache,
+) (bool, error) {
+	strategy, err := vm.RunStrategy()
+	if err != nil {
+		return false, fmt.Errorf("error getting run strategy for vm %s in namespace %s: %v", vm.Name, vm.Namespace, err)
+	}
+
+	// VM is stopped from GUI
+	if strategy == kubevirtv1.RunStrategyHalted {
+		return true, nil
+	}
+
+	if vm.Status.PrintableStatus != kubevirtv1.VirtualMachineStatusStopped {
+		return false, nil
+	}
+
+	// When vm is stopped inside VM, the vmi will not be deleted.
+	// The status of vmi will be "Succeeded".
+	vmi, err := vmiCache.Get(vm.Namespace, vm.Name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return false, fmt.Errorf("error getting vmi %s in namespace %s: %v", vm.Name, vm.Namespace, err)
+	}
+
+	if vmi == nil {
+		logrus.Debugf("VM %s/%s is deleted", vm.Namespace, vm.Name)
+		return true, nil
+	}
+
+	if vmi.IsFinal() {
```

Comment:
@Yu-Jack, `IsFinal` also checks whether VMI is `Failed`. Do you know when will VMI gets `Failed` and whether KubeVirt controller restarts it? Thanks.

Reply:
I'm not sure what I'm looking for is right place, but it seems KubeVirt doesn't support start failed state vm. 
- virt-api
  - https://github.com/kubevirt/kubevirt/blob/v0.54.0/pkg/virt-api/rest/subresource.go#L535-L538
- virt-controller
  - https://github.com/kubevirt/kubevirt/blob/v0.54.0/pkg/virt-controller/watch/vmi.go#L556-L574
  - https://github.com/kubevirt/kubevirt/blob/v0.54.0/pkg/virt-controller/watch/vmi.go#L986-L992

In general, it says `Failed mean the vmi crashed, disappeared unexpectedly or got deleted from the cluster before it was ever started`. For one example, the vmi is created, but is not still scheduled. If it's deleted, it will become Failed state in https://github.com/kubevirt/kubevirt/blob/v0.54.0/pkg/virt-controller/watch/vmi.go#L435.

Reply:
I found that we can make VMI as Failed if we update `kubevirt.io/heartbeat` annotation  to an error format in Node. In this case, VM is running but VMI is failed. However, this is an edge case. I think we can just skip it. I am good with `IsFinal` here. Thanks.

Ref: https://github.com/kubevirt/kubevirt/blob/e8e638edc22587ec7be2cc3d983b61763e33f973/pkg/virt-controller/watch/node.go#L186-L203